File: Graphs - DFS/547. Number of Provinces/solution.ts
function findCircleNum(isConnected: number[][]): number {
    function dfs(i: number) {
        for (let j = 0; j < isConnected.length; j++) {
            if (isConnected[i][j] === 1) {
                isConnected[i][j] = isConnected[j][i] = 0;
                dfs(j);
            }
        }
    }

    let provinces = 0;
    for (let i = 0; i < isConnected.length; i++) {
        if (isConnected[i][i] === 1) {
            provinces++;
            dfs(i);
        }
    }
    return provinces;
};

File: Graphs - DFS/547. Number of Provinces/solution.rs
impl Solution {
    pub fn find_circle_num(is_connected: Vec<Vec<i32>>) -> i32 {
        let mut is_connected = is_connected;
        let mut provinces = 0;
        for i in 0..is_connected.len() {
            if is_connected[i][i] == 1 {
                provinces += 1;
                Self::dfs(i, &mut is_connected);
            }
        }
        provinces
    }

    fn dfs(i: usize, is_connected: &mut Vec<Vec<i32>>) {
        for j in 0..is_connected.len() {
            if is_connected[i][j] == 1 {
                is_connected[i][j] = 0;
                is_connected[j][i] = 0;
                Self::dfs(j, is_connected);
            }
        }
    }
}